### PR TITLE
feat: Compiler optimization: Omit wrappers around single other type

### PR DESCRIPTION
**Pull request by RustanLeino**
This PR introduces a compiler optimization for datatypes. In particular,

* A type
    ```` dafny
    datatype Record = Record(x: int)
    ````
    is compiled to the target type `int`. More generally, if an inductive `datatype` has exactly one non-ghost constructor and that constructor has exactly one non-ghost member (and some other conditions apply, see fine print below), then values of the datatype are compiled as that member, without the enclosing datatype-constructor "wrapper".
* A tuple type with ghost components is compiled to the smaller tuple without the ghost components. For example, the 4-tuple type `(ghost int, real, ghost bool, bv3)` is compiled to the 2-tuple type `(real, bv3)`. (This optimization is not entirely new with this PR, but this PR makes the treatment of this more uniform.)
* As a special case of the above, if a ghost tuple is trimmed down to a non-ghost 1-tuple, then (on behalf of the first bullet above), the target type is further changed to be just that non-ghost component. For example, `(int, ghost int)` is compiled as just `int`.

## Benefits

The optimization has several benefits. As one example, a function

``` dafny
function F(x: X): (Result, ghost AuxiliaryData)
```

is compiled to just

``` dafny
function F(x: X): Result
```

That is, there is no run-time cost of having also returned a ghost component in the Dafny source code.

As a second example, types that use datatype constructors as "tags" get optimized in the compiled code. For instance,

``` dafny
datatype Box<X> = Box(value: X)
```

is compiled to just `X` (whatever `X` happens to be).

As a third example, types with uninitialized values (using ghost constructors) have any run-time cost. For instance,

``` dafny
datatype MaybeInitialized<X> =
  | ghost Uninitialized
  | Initialized(value: X)
```

is compiled to just `X` (and the uninitialized value is typically compiled to a zero-equivalent value of the target type `X`).

## The fine print

The optimization is _not_ performed when one of the following applies:

* the datatype is declared with `{:extern}`
* the datatype is a `codatatype`
* the datatype has a `const` member
* the datatype declaration is involved in a cycle; for example,
    ``` dafny
    datatype MyselfForever = ghost Nothing | Something(MyselfForever)
    ```
    is not optimized
* the command-line option `/optimizeErasableDatatypeWrapper:0` is used.

## Supporting optimizations

The PR provides full compilation support for C#, Java, JavaScript, Go, and Python (but not C++, since the Dafny-to-C++ compiler lacks many features). (In the implementation, the compiler getter `OptimizesErasableDatatypeWrappers` lets a compiler, like C++, opt out of the optimization altogether.)

Destructors (and update expressions) are optimized accordingly. For example, if `r` has type `Record` from above, then `r.x` is compiled into (the target code for) `r` and `r.(x := E)` is compiled as just `E`. As another example, if `v` has type `(ghost int, int, int)`, then `v.1` is compiled into `v.0`.

For a datatype `T` with exactly one non-ghost constructor, say `C`, a discriminator expression `t.C?` is compiled as just `true`.

The PR also optimizes an expression `d.(g := E)` into just `d` if `g` is a ghost field.

Other things that the optimization requires (and this PR handles) are:

* Equality. For example, if `r` and `s` are of type `Record` above, then the `Record` comparison `r == s` is compiled into the integer comparison `r.x == s.x`.
* `match` constructs.
* Default values.
* Type descriptors.
* For C#, `DowncastClone`.
* Instance members of optimized types turn into custom-receiver static members.

## Compatibility

For code written entirely in Dafny, this PR should have no impact (other than optimized behavior at run time). Code that links to the Dafny code directly will notice the missing wrappers. Such code can add the `{:extern}` attribute to the `datatype` declaration, since that will disable the optimization. To opt out of the optimization entirely, use `/optimizeErasableDatatypeWrapper:0`.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/Compilers/Compiler-python.cs
Comment (fabiomadge) [inline Source/DafnyCore/Compilers/Compiler-python.cs:1349]:
```suggestion
      wr.Write(source);
      if (IsInvisibleWrapper(ctor.EnclosingDatatype, out var coreDtor)) {
        Contract.Assert(coreDtor.CorrespondingFormals.Count == 1);
        Contract.Assert(dtor == coreDtor.CorrespondingFormals[0]); // any other destructor is a ghost
      } else {
        wr.Write(ctor.EnclosingDatatype is TupleTypeDecl ? $"[{dtor.NameForCompilation}]" : $".{IdProtect(dtor.CompileName)}");
      }
```

File: Source/DafnyCore/Compilers/Compiler-python.cs
Comment (fabiomadge) [inline Source/DafnyCore/Compilers/Compiler-python.cs:756]: Can we move this inside the else block?

File: Source/DafnyCore/Resolver.cs
Comment (RustanLeino) [reply Source/DafnyCore/Resolver.cs:14697]: Good point. I had placed it there to be next to the related `SubstType` method, but it is equally out of place.

I create PR https://github.com/dafny-lang/dafny/pull/3057 to move two previous type-related utility methods into a new place: a class `TypeUtil`. I will merge with that PR and will move `ReplaceTypeArguments` there as well.

**Comment by RustanLeino** [comment]
> Could this optimisation be done in a pass that runs before SinglePassCompiler?
> 
> I think we agreed as a group that transformations which can be implemented as separate passes, should be. The benefit in this case is that we wouldn't add more code to a transformation that's already complex.

@keyboardDrummer Yes, that would be better and cleaner! (I had started this PR over a year ago and had returned to it now to finish it up.) To do this, we need some new machinery for cloning/substituting resolved expressions/statements/declarations. Once we have that machinery, I imagine we'll use it a lot.

Should we put this PR on hold (indefinitely) and reimplement it as a transformation once we (separately) build that machinery?

File: Source/DafnyCore/Compilers/Compiler-python.cs
Comment (RustanLeino) [reply Source/DafnyCore/Compilers/Compiler-python.cs:756]: No, the updated `type` is also needed in the `if` guard.

Comment (fabiomadge) [reply Source/DafnyCore/Compilers/Compiler-python.cs:756]: Are you sure?

**Comment by keyboardDrummer** [comment]
> @keyboardDrummer Yes, that would be better and cleaner! (I had started this PR over a year ago and had returned to it now to finish it up.) To do this, we need some new machinery for cloning/substituting resolved expressions/statements/declarations. Once we have that machinery, I imagine we'll use it a lot.

Note that we're adding the ability to clone resolved AST nodes in https://github.com/dafny-lang/dafny/pull/2734

> Should we put this PR on hold (indefinitely) and reimplement it as a transformation once we (separately) build that machinery?

I'm OK with both. I think this PR adds a small complexity increase to the compilation back-ends. If there's no rush in merging this then I think waiting until https://github.com/dafny-lang/dafny/pull/2734 lands is better, but if a user would greatly benefit from this now then let's merge it.

File: Source/DafnyCore/Compilers/Compiler-python.cs
Comment (fabiomadge) [inline Source/DafnyCore/Compilers/Compiler-python.cs:275]: We only use `simplifiedType` here, but we don't even need it, because `ForcePlaceboValue` does the same step for us.
```suggestion
        wDefault.WriteLine($"return lambda: {ForcePlaceboValue(UserDefinedType.FromTopLevelDecl(dt.tok, dt), wDefault, dt.tok)}");
```

File: Source/DafnyCore/Compilers/Compiler-python.cs
Comment (fabiomadge) [inline Source/DafnyCore/Compilers/Compiler-python.cs:757]:
```suggestion
      if (type.NormalizeExpandKeepConstraints() is UserDefinedType {ResolvedClass: DatatypeDecl dt} udt && IsInvisibleWrapper(dt, out _)) {
```

File: Source/DafnyCore/Compilers/Compiler-python.cs
Comment (fabiomadge) [inline Source/DafnyCore/Compilers/Compiler-python.cs:1183]:
```suggestion
      switch (GetMemberStatus(member)) {
        case MemberCompileStatus.Identity:
          return SimpleLvalue(obj);
        case MemberCompileStatus.AlwaysTrue:
          return StringLvalue("True");
```

**Review by fabiomadge — COMMENTED** [review]
Sorry for revisiting.

**Comment by RustanLeino** [comment]
> > Should we put this PR on hold (indefinitely) and reimplement it as a transformation once we (separately) build that machinery?
>
> I'm OK with both. I think this PR adds a small complexity increase to the code-generators. If there's no rush in merging this then I think waiting until https://github.com/dafny-lang/dafny/pull/2734 lands is better, but if a user would greatly benefit from this now then let's merge it.

@keyboardDrummer Thinking about this some more, I suggest we then work toward merging this PR as is. This will add the feature and also includes some other fix, and will remove the need to constantly keep this PR is limbo and up-to-date with the latest sources. Then, when we're ready to start writing transforms for our compiler, we can use this as an early application thereof. The `IsInvisibleWrapper` and `SimplifyType` methods in this PR will be used by the transformation, and so will the test cases added in this PR.

**Comment by keyboardDrummer** [comment]
> > > Should we put this PR on hold (indefinitely) and reimplement it as a transformation once we (separately) build that machinery?
> > 
> > 
> > I'm OK with both. I think this PR adds a small complexity increase to the code-generators. If there's no rush in merging this then I think waiting until #2734 lands is better, but if a user would greatly benefit from this now then let's merge it.
> 
> @keyboardDrummer Thinking about this some more, I suggest we then work toward merging this PR as is. This will add the feature and also includes some other fix, and will remove the need to constantly keep this PR is limbo and up-to-date with the latest sources. Then, when we're ready to start writing transforms for our compiler, we can use this as an early application thereof. The `IsInvisibleWrapper` and `SimplifyType` methods in this PR will be used by the transformation, and so will the test cases added in this PR.

Sounds good

File: Source/DafnyCore/Compilers/Compiler-python.cs
Comment (fabiomadge) [inline Source/DafnyCore/Compilers/Compiler-python.cs:1181]:
```suggestion
          return StringLvalue(True);
```

File: Source/DafnyCore/Compilers/SinglePassCompiler.cs
Comment (keyboardDrummer) [inline Source/DafnyCore/Compilers/SinglePassCompiler.cs:1088]: I'm not sure I understand the purpose of this check. Why can we not always optimise away the constructor at this point?

If I have

```dafny
datatype Wrapper = Wrapper(ints: Ints)
datatype Ints = Ints(x: int, y: int) | AnotherWrapper(wrapper: Wrapper)
```

where it seems `properlyContainedTypeDecls.Contains(dt) == true`, then wouldn't we still be able to write that as

```dafny
datatype Ints = Ints(x: int, y: int) | AnotherWrapper(wrapper: Ints)
```

?

File: Source/DafnyCore/Compilers/SinglePassCompiler.cs
Comment (keyboardDrummer) [inline Source/DafnyCore/Compilers/SinglePassCompiler.cs:1068]: What does `invisible` in the name mean? Could we rename this to `ShouldDatatypeBeErased` ?

File: Source/DafnyCore/DafnyOptions.cs
Comment (keyboardDrummer) [inline Source/DafnyCore/DafnyOptions.cs:430]: If you want this option to be available in the new CLI UI (and you do), then you have to add a value of `IOptionSpec` to `ICommandSpec.ExecutionOptions`. You can add the same value to `DafnyOptions.AvailableNewStyleOptions` to also make it available in the old UI.

File: Source/DafnyCore/Compilers/SinglePassCompiler.cs
Comment (keyboardDrummer) [inline Source/DafnyCore/Compilers/SinglePassCompiler.cs:1133]: `typle`

Comment (keyboardDrummer) [inline Source/DafnyCore/Compilers/SinglePassCompiler.cs:1133]: I'm quite looking forward to seeing this be implemented as a separate pass. I think it'll improve the readability a lot.

File: Source/DafnyCore/Compilers/SinglePassCompiler.cs
Comment (keyboardDrummer) [inline Source/DafnyCore/Compilers/SinglePassCompiler.cs:1054]: I would move the new members of `SinglePassCompiler`, GetMemberStatus, IsInvisibleWrapper, CompiledTypes and SimplifyType to a separate class with a name like `WrapperTypeEraser`.

File: Source/DafnyCore/AST/TypeUtil.cs
Comment (RustanLeino) [reply Source/DafnyCore/AST/TypeUtil.cs:9]: Oh my! Yes, this is the result of a bad merge. (I had merged the other PR into this one before the other one was merged into master.)

File: Source/DafnyCore/AST/TypeUtil.cs
Comment (RustanLeino) [reply Source/DafnyCore/AST/TypeUtil.cs:23]: Yep, sorry. Bad merge.

File: Source/DafnyCore/AST/TupleTypeDecl.cs
Comment (RustanLeino) [reply Source/DafnyCore/AST/TupleTypeDecl.cs:16]: I made `NonGhostTupleTypeDecl` into a property getter.

I don't understand what you mean to say with the `RuntimeVersion` code snippet.

File: Source/DafnyCore/AST/BuiltIns.cs
Comment (RustanLeino) [reply Source/DafnyCore/AST/BuiltIns.cs:239]: It will just generate a `Tuple1` class as well. That seems alright. I'll change the test to just `nonGhostDims != dims`.

File: Source/DafnyCore/Compilers/Compiler-Csharp.cs
Comment (RustanLeino) [reply Source/DafnyCore/Compilers/Compiler-Csharp.cs:307]: I added a comment, referring to `docs/Compilation/AutoInitialization.md`.

File: Source/DafnyCore/Compilers/SinglePassCompiler.cs
Comment (RustanLeino) [reply Source/DafnyCore/Compilers/SinglePassCompiler.cs:1069]: I added some comments in the doc-header of the method.

File: Source/DafnyCore/Compilers/SinglePassCompiler.cs
Comment (RustanLeino) [reply Source/DafnyCore/Compilers/SinglePassCompiler.cs:1068]: That would be a possibility, but I chose the bool-return/out-parameter idiom instead.

File: Source/DafnyCore/Compilers/SinglePassCompiler.cs
Comment (RustanLeino) [reply Source/DafnyCore/Compilers/SinglePassCompiler.cs:1079]: My new doc-comment about the method address this now.

File: Source/DafnyCore/Compilers/SinglePassCompiler.cs
Comment (RustanLeino) [reply Source/DafnyCore/Compilers/SinglePassCompiler.cs:1087]: I renamed it to `GetCompiledComponentTypes` and improve its doc-comment.

File: Source/DafnyCore/Compilers/SinglePassCompiler.cs
Comment (RustanLeino) [reply Source/DafnyCore/Compilers/SinglePassCompiler.cs:1088]: Very good catch! The current algorithm is too strict and misses the opportunity to optimize your `Wrapper` type. I'll improve the algorithm.

As for the question about the purpose of the check, here is an answer: If optimizing away the constructor gives back the type itself, then there would be a problem. To illustrate, it may help to imagine a type like

``` dafny
datatype Cycle = Ctor(Cycle)
```

It would not be possible to remove the wrapper `Ctor`, because doing so would give another type whose wrapper `Ctor` we'd remove, and it is also a wrapper `Ctor` that we'd remove, etc. However, Dafny doesn't actually allow an `datatype` like `Cycle`, so to illustrate the real problem, I have to show a more complicated type. Here is one:

``` dafny
datatype GList =
  | ghost End
  | More(GList)
```

Dafny does allow this type. We don't represent `End` at run time, since it's ghost. The other constructor has one argument, so it looks like it could be optimized. But this would then run into the infinite-cycle problem I gave a taste of with the `Cycle` type above.

File: Source/DafnyCore/Compilers/SinglePassCompiler.cs
Comment (RustanLeino) [reply Source/DafnyCore/Compilers/SinglePassCompiler.cs:1068]: With "invisible wrapper", I meant that the single compiled constructor is "erasable". I considered renaming it `IsEraseableDatatype`, but is a bit misleading, since it's only the "wrapper" part that's erased--not the entire datatype value. I'll rename it `IsEraseableDatatypeWrapper` and will do the same renaming for the test file and command-line option.

File: Source/DafnyCore/Compilers/SinglePassCompiler.cs
Comment (RustanLeino) [reply Source/DafnyCore/Compilers/SinglePassCompiler.cs:1054]: I moved them. The way I did it, the qualified names now contain some redundancy.

File: Source/DafnyCore/DafnyOptions.cs
Comment (RustanLeino) [reply Source/DafnyCore/DafnyOptions.cs:432]: Indeed, I do. I changed it. Well, I don't understand if the `Description` should talk about arguments `0` and `1` or arguments `false` and `true`.

File: Source/DafnyCore/AST/BuiltIns.cs
Comment (RustanLeino) [reply Source/DafnyCore/AST/BuiltIns.cs]: The suggested change worked. However, in the end, I decided to revert it to the code above. That is, the test does include `nonGhostDims != 1 || !DafnyOptions.O.Compiler.OptimizesInvisibleDatatypeWrappers` again, so that we will _not_ generate a `Tuple1` class. It would seem odd to generate a class that we know is never going to be used.

File: Source/DafnyCore/DafnyOptions.cs
Comment (keyboardDrummer) [reply Source/DafnyCore/DafnyOptions.cs]: > Well, I don't understand if the Description should talk about arguments 0 and 1 or arguments false and true

Yes that's a problem. I'm adding the option of having two descriptions here: https://github.com/dafny-lang/dafny/pull/3154

**Review by keyboardDrummer — APPROVED** [review]

File: Source/DafnyCore/Compilers/DatatypeWrapperEraser.cs
Comment (RustanLeino) [reply Source/DafnyCore/Compilers/DatatypeWrapperEraser.cs:164]: Nice suggestion. I'll be sure to do that when I have ISet's in the future.

File: Source/DafnyCore/Compilers/SinglePassCompiler.cs
Comment (RustanLeino) [reply Source/DafnyCore/Compilers/SinglePassCompiler.cs]: I thought, in my old-school brain, that the indices were clear enough for this task. Your suggested names would be much better than my `i` and `j`.

Yes, it was the creation of a new `List` that I was avoiding. Then again, I imagine the creation of the various lambda expressions (like `d => !d.IsGhost`) will take as much time and space as an additional `List`, so not much is saved. Because I used `firstDestructorIndex + 1` as the starting point, I believe the list would be traversed only once, but unless the C# compiler is good at optimizing, it would create a lambda expression for each invocation of `FindIndex`.

Regarding copying the entire AST and omitting ghost things: I'm very much in favor of this. A lot of the code that is duplicated across the compilers has to do with avoiding ghosts, so that would be so nice to do once and for all.

**Pull request merged**
